### PR TITLE
Add tests to ensure that we don't break `current_time`'s underlying driver

### PR DESCRIPTION
**Pull request by MabezDev**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

### Pull Request Details 📖

Adds tests to ensure we don't break invariants required for `current_time` to work. It also pulls `esp-metadata` into HIL test, so we can use the nice cfgs we have in esp-hal to nicely cfg away specific test cases.

### CODE REVIEW

File: esp-hal/src/timer/systimer.rs [hunk 1]
```diff
@@ -131,6 +132,9 @@ impl<'d> SystemTimer<'d> {
         #[cfg(soc_etm)]
         etm::enable_etm();
 
+        // Don't reset Systimer as it will break `current_time`, only enable it
+        PeripheralClockControl::enable(PeripheralEnable::Systimer);
```
Comment (Dominaezzz) [inline esp-hal/src/timer/systimer.rs:136]: Looking at the C6's TRM, this should be done before the `etm::enable_etm();` call above.

Since this peripheral may still be in reset, the register writes in that method will be ignored.

Comment (MabezDev) [reply esp-hal/src/timer/systimer.rs:133]: Good catch, thanks!

**Review by jessebraham — APPROVED** [review]
Thanks!

File: hil-test/Cargo.toml [hunk 2]
```diff
@@ -227,3 +231,6 @@ incremental      = false
 opt-level        = 3
 lto              = "fat"
 overflow-checks  = false
+
+[lints.rust]
+unexpected_cfgs = "allow"
```
Comment (bugadani) [inline hil-test/Cargo.toml:236]: Why is this necessary?

File: hil-test/tests/get_time.rs [hunk 3]
```diff
@@ -45,4 +53,37 @@ mod tests {
         assert!(t2 > t1);
         assert!((t2 - t1).to_millis() >= 500u64);
     }
+
+    #[cfg(systimer)]
+    #[test]
+    #[timeout(3)]
+    fn test_current_time_construct_systimer(ctx: Context) {
+        let t1 = esp_hal::time::current_time();
```
Comment (bugadani) [inline hil-test/tests/get_time.rs:61]: What does this test test, actually? If it's only "we don't reset current_time", then the delay is redundant and all you want is "t2 >= t1". Other than that, we have plenty of "delay 500ms and assert that we really delayed for 500ms" tests elsewhere IMO and this just makes CI take half a second longer.

I would also not duplicate the test function, as only a single statement is different between them.

**Review by bugadani — COMMENTED** [review]
I'm not sure if the PR title is actually in line with the actual changes (the PR is more about resetting systimer at boot time)

File: hil-test/Cargo.toml (hunk 2 above)
Comment (MabezDev) [reply hil-test/Cargo.toml:236]: Because esp-meta emit's cfgs that rust doesn't know about in build.rs

Comment (bugadani) [reply hil-test/Cargo.toml:236]: In that case I'd prefer unlinting this in build.rs and not crate-wide, I think.

Comment (MabezDev) [reply hil-test/Cargo.toml:236]: Sorry, I'm not being clear. It emits cfgs for the entire crate from build.rs. We have the same allow rule in esp-hal.

Comment (bugadani) [reply hil-test/Cargo.toml:236]: Right, that makes it clearer, thanks

File: hil-test/tests/get_time.rs (hunk 3 above)
Comment (bjoernQ) [reply hil-test/tests/get_time.rs:61]: 500 millis is probably a bit long here but if we have a delay shouldn't we delay before we construct the timer?

File: hil-test/tests/get_time.rs [hunk 4]
```diff
@@ -20,10 +24,22 @@ impl Context {
...
+        }
     }
 }
 
+fn time_moves_forward_during<F: FnOnce(Context)>(ctx: Context, f: F) {
+    let t1 = esp_hal::time::current_time();
+    f(ctx);
+    let t2 = esp_hal::time::current_time();
+
+    assert!(t2 > t1);
```
Comment (bugadani) [inline hil-test/tests/get_time.rs:40]: Can the CPUs be so fast we end up with t2 == t1?

**Review by bugadani — APPROVED** [review]

File: hil-test/tests/get_time.rs (hunk 4 above)
Comment (MabezDev) [reply hil-test/tests/get_time.rs:40]: I don't think so, no
